beverage-machine/src/com/tarktech/training/beverage/BeverageMachine.java

```diff
@@ -1,33 +1,63 @@
-        return beverageMaker.dispenseBeverage();
-    }
-
-    public int getDispensedBeverageCount(BeverageType beverageType) {
-        BeverageMaker beverageMaker = beverageMakers.get(beverageType);
-        return beverageMaker.getDispensedBeverageCount();
-    }
-
-    public int getAvailableBeverageCount(BeverageType beverageType) {
-        BeverageMaker beverageMaker = beverageMakers.get(beverageType);
-        return beverageMaker.getAvailableBeverageCount();
-    }
-}
+package com.tarktech.training.beverage;
+
+
+import com.sun.source.tree.ReturnTree;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.tarktech.training.beverage.BeverageType.*;
+
+public class BeverageMachine {
+    private Map<BeverageType, BeverageMaker> beverageMakers = new HashMap<>();
+
+    public BeverageMachine() {
+        beverageMakers.put(Latte, new LatteMaker());
+        beverageMakers.put(Tea, new TeaMaker());
+        beverageMakers.put(Cappuccino, new CappuccinoMaker());
+        beverageMakers.put(Mocha, new MochaMaker());
+    }
+
+    public String dispenseBeverage(BeverageType beverageType) {
+        BeverageMaker beverageMaker = beverageMakers.get(beverageType);
+        return beverageMaker.dispenseBeverage();
+    }
+
+    public int getDispensedBeverageCount(BeverageType beverageType) {
+        BeverageMaker beverageMaker = beverageMakers.get(beverageType);
+        return beverageMaker.getDispensedBeverageCount();
+    }
+    public int getAvailableBeverageCount(BeverageType beverageType) {
+        BeverageMaker beverageMaker = beverageMakers.get(beverageType);
+        return beverageMaker.getAvailableBeverageCount();
+    }
+    public String getMostFavouriteBeverageCount(BeverageType beverageType) {
+
+        if(beverageMakers.get(Tea).getDispensedBeverageCount()>beverageMakers.get(Latte).getDispensedBeverageCount()){
+            if(beverageMakers.get(Tea).getDispensedBeverageCount()>beverageMakers.get(Mocha).getDispensedBeverageCount()) {
+                if (beverageMakers.get(Tea).getDispensedBeverageCount() > beverageMakers.get(Cappuccino).getDispensedBeverageCount()) {
+                return "Tea";
+
+                }
+                return "Cappucino";
+
+            }
+            return "Mocha";
+        }
+        return "Latte";
+    }
+    public String getLeastFavouriteBeverageCount(BeverageType beverageType) {
+        if(beverageMakers.get(Tea).getDispensedBeverageCount()<beverageMakers.get(Latte).getDispensedBeverageCount()){
+            if(beverageMakers.get(Tea).getDispensedBeverageCount()<beverageMakers.get(Mocha).getDispensedBeverageCount()) {
+                if (beverageMakers.get(Tea).getDispensedBeverageCount() < beverageMakers.get(Cappuccino).getDispensedBeverageCount()) {
+                    return "Tea";
+
+                }
+                return "Cappucino";
+
+            }
+            return "Mocha";
+        }
+        return "Latte";
+    }
+}

```

Comment:
- We don't require the **"beverageType"** argument in most/least favorite beverage method, as explained earlier.
- When naming **variables/methods**, try to use **camel case**.
- Use meaningful names for variables/methods like **"getMostFavoriteBeverage"**, **"getLeastFavoriteBeverage"**, etc.
- For the time being, using "if" to determine the most favorite beverage is a good strategy, however consider the following example:
- You need to add 5 more beverage type:
1. Black Coffee
2. Espresso
3. Iced Coffee
4. Iced Latte
5. Macchiato

- **More nested if statements** will be required now. As a result, you probably break the functionality.
- Instead you may follow below pseudo code:
1. You can get list of enum values using **Enum.values()** or equivalent library function.
2. Use an algorithm, similar to find **min/max from an array** to identify most/least favorite beverage.